clm/model_utils_gpu.py

```diff
@@ -0,0 +1,152 @@
+import os
+from omegaconf.omegaconf import OmegaConf, open_dict
+import torch
+
+from nemo.utils import logging
+from nemo.collections.nlp.data.language_modeling.megatron.gpt_sft_chat_dataset import get_prompt_template_example
+
+from nemo_aligner.utils.train_script_utils import (
+    init_distributed,
+    init_peft,
+)
+from nemo_aligner.utils.utils import load_from_nemo
+from nemo_aligner.models.nlp.gpt.gpt_sft_model import GPTSFTModel
+
+
+def setup_distributed(config):
+    """Initialize torch.distributed."""
+    # Get rank and world size.
+    local_rank = int(os.getenv("LOCAL_RANK", 0))
+    rank = int(os.getenv('RANK', '0'))
+    world_size = int(os.getenv("WORLD_SIZE", '1'))
+
+    logging.info(
+        f'Initializing torch.distributed with local_rank: {local_rank}, rank: {rank}, world_size: {world_size}'
+    )
+
+    # Set the device id.
+    device = rank % torch.cuda.device_count()
+    if local_rank is not None:
+        device = local_rank
+    torch.cuda.set_device(device)
+
+    # Call the init process.
+    init_method = 'tcp://'
+    master_ip = os.getenv('MASTER_ADDR', 'localhost')
+    master_port = os.getenv('MASTER_PORT', '6000')
+    import datetime
+    DEFAULT_TIMEOUT = datetime.timedelta(minutes=60)
+    init_method += master_ip + ':' + master_port
+    torch.distributed.init_process_group(backend='nccl', timeout=DEFAULT_TIMEOUT, world_size=world_size, rank=rank, init_method=init_method)
+    return local_rank, rank, world_size
+
+def setup_model_optimizer(config, trainer):
+    logging.info("loading model")
+    model, updated_cfg = load_from_nemo(
```

Comment:
Looks good to me if we need `load_from_nemo`.

Alternatively, if we can load from state_dict directly consider the following alternative`:
```
CustomMegatronGPTModel(MegatronGPTModel):
    def load_model_state_dict(self, checkpoint) -> None:
        ...
```
[here](https://github.com/mlcommons/training_results_v3.1/blob/main/NVIDIA/benchmarks/gpt3/implementations/pytorch/custom_callbacks.py#L503-L517)